rfd/0048-desktop-access-session-recording.md

```diff
@@ -0,0 +1,300 @@
+ultimately writing them to persistent storage (a file or S3-compatible storage).
+
+In asynchronous recording mode, we also leverage a
+[`TeeStreamer`](https://github.com/gravitational/teleport/blob/bfe7f9878a05bce08129ce92ee8773bace7fd75b/lib/srv/sess.go#L1074)
+which is responsible for filtering out `SessionPrint` events that should not be
+written to Teleport's main audit log.
+
+#### Playback
+
+In order to maintain compatibility with older versions of Teleport, SSH session
+playback converts this stream of `SessionPrint` events into a legacy format
+consisting of two files:
+
+- an "events" file containing timing data
+- a "chunks" file containing the raw PTY data
+
+For SSH, this format has the nice property of the "chunks" file being useful on
+its own, as the raw data can be viewed and searched independent of the timing
+data (outside of the Teleport player). This separation of timing data is not
+necessary for desktop sessions, as the raw data cannot be interpreted outside of
+the Teleport player, so this conversion step will not be necessary for desktop
+sessions.
+
+Teleport's existing API also exposes an endpoint for streaming audit events associated
+with a particular session.
+
+```
+func StreamSessionEvents(
+    ctx context.Context,
+    sessionID string,
+    startIndex int64) (chan events.AuditEvent chan error)
+```
+
+On the backend, the log is downloaded from external storage, the events are
+parsed via an `AuditReader`, and finally written to the events channel. There is
+no attention paid to timing at this step. Events are written to the channel as
+soon as they are available and the process is naturally throttled by how fast
+the consumer reads from the channel.
+
+### Overview
+
+During a live session, desktop images are sent to the browser via a series of
+PNG frames encoded in a TDP message (message #2, "PNG frame"). At a high level,
+the session recording feature needs to archive these messages in the audit log
+so that they can be played back at a later time.
+
+During playback, the majority of the frontend code works exactly as it does
+during a live session - the PNGs are rendered on an HTML canvas.
+
+The major differences in playback mode are:
+
+- no user-input is captured or sent across the wire (mouse clicks, scroll wheel, etc)
+- playback features such as play/pause (at a minimum), and seek are desired
+
+#### Compatibility
+
+The session recording feature further increases the importance of
+backwards-compatible changes to the TDP spec, as newer versions of Teleport must
+be able to play back recordings captured by older versions.
+
+This means we must favor adding new messages over modifying existing messages,
+and retain support for deprecated messages in order to keep recordings valid.
+
+#### Recording
+
+Teleport session recordings are stored as an ordered sequence of protobuf-encoded
+audit events. Desktop session recording will work in the same way.
```

Comment:
Do we planing to introduce some per session DestopRecordingEvent   that will collect some windows desktop session data like session length ?

Reply:
Added more detail here. We'll take the same approach used by SSH sessions, which is that the `session.end` audit event has enough information (session ID, start time, end time, etc) for the UI to display session metadata.